h/api/test/auth_test.py

```diff
@@ -28,3 +31,209 @@ def test_translate_annotation_principals(p_in, p_out):
+def test_userid_from_bearer_token_passes_audience_to_decode(decode):
+    request = mock_request(token='Bearer abc123')
+
+    auth.userid_from_bearer_token(request)
+
+    assert decode.call_args[1]['audience'] == request.host_url
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_passes_leeway_to_decode(decode):
+    request = mock_request(token='Bearer abc123')
+
+    auth.userid_from_bearer_token(request)
+
+    assert decode.call_args[1]['leeway'] == 240
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_passes_algorithm_to_decode(decode):
+    request = mock_request(token='Bearer abc123')
+
+    auth.userid_from_bearer_token(request)
+
+    assert decode.call_args[1]['algorithm'] == 'HS256'
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_returns_sub_from_decode(decode):
+    decode.return_value = {'sub': 'acct:test_user@hypothes.is'}
+
+    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+
+    assert result == 'acct:test_user@hypothes.is'
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_returns_None_if_no_sub(decode):
+    decode.return_value = {}  # No 'sub' key.
+
+    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+
+    assert result is None
+
+
+@user_id_from_bearer_token_fixtures
+def test_userid_from_bearer_token_returns_None_if_decoding_fails(decode):
+    decode.side_effect = jwt.InvalidTokenError
+
+    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+
+    assert result is None
+
+
+unauthenticated_userid_from_bearer_token_fixtures = pytest.mark.usefixtures(
+    'decode')
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_no_Authorization():
+    request = mock_request()
+    request.headers = {}
+
+    result = auth.unauthenticated_userid_from_bearer_token(request)
+
+    assert result is None
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_calls_decode(decode):
+    request = mock_request(token='Bearer abc123')
+
+    auth.unauthenticated_userid_from_bearer_token(request)
+
+    decode.assert_called_once_with('abc123', verify=False)
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_returns_userid(decode):
+    request = mock_request(token='Bearer abc123')
+    decode.return_value = {'sub': 'acct:test_user@hypothes.is'}
+
+    result = auth.unauthenticated_userid_from_bearer_token(request)
+
+    assert result == 'acct:test_user@hypothes.is'
+
+
+@unauthenticated_userid_from_bearer_token_fixtures
+def test_unauthenticated_userid_from_bearer_token_when_decode_fails(decode):
+    request = mock_request(token='Bearer abc123')
+    decode.side_effect = jwt.InvalidTokenError
+
+    result = auth.unauthenticated_userid_from_bearer_token(request)
+
+    assert result is None
+
```

Comment:
Yeah, I'll add that. Actually originally with all the `encode()` tests, instead of testing that `encode()` was called with the right 'sub' for example, I would take the result of `encode()` and decode it (by calling jwt directly in the tests) and test that it decoded to the right sub.

Similarly with the `decode()` tests, I called jwt directly to encode something and pass it in.

In the way that seemed like a better way to test this to me as it tests that it really succeeds and fails when we think it does. But it's not how most of our unit tests work. and it was testing the jwt lib.

I still feel a bit unsure about this kind of "it should call library foo with the params that I think it should" kind of tests. Yes, but what if I misread the foo docs and the params that I think it should be using don't actually cause the foo lib to do what I think it does?

Anyway, I'll add the one invariant test.